packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,328 @@
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.none) {
+        return '#CCC';
+      } else {
+        return state === State.valid ? '#13AA52' : '#CF4A22';
+      }
+    }
+
+    function getInputPaddingFromState() {
+      if (state === State.valid || state === State.error) {
+        return '30px';
+      } else {
+        return optional ? '60px' : '12px';
+      }
+    }
+
+    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
+      if (onChange != undefined) {
+        state = onChange(e.target.value);
+      }
+
+      if (setValue != undefined) {
+        setValue(e.target.value);
+      }
+    }
+
+    const interactionRing = css`
+      transition: all 150ms ease-in-out;
+      transform: scale(0.9, 0.8);
+      border-radius: 7px;
+      position: absolute;
+      top: -3px;
+      bottom: -3px;
+      left: -3px;
+      right: -3px;
+      pointer-events: none;
+      background-color: ${uiColors.gray.light2};
+    `;
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const textInputStyle = css`
+      display: flex;
+      flex-direction: column;
+    `;
+
+    const labelStyle = css`
+      color: #3d4f58;
+      font-size: 14px;
+      font-weight: bold;
+      height: 20px;
+      padding-bottom: 8px;
+    `;
+
+    const descriptionStyle = css`
+      color: #5d6c74;
+      font-size: 14px;
+      height: ${description === '' ? '0' : '20'}px;
+      font-weight: normal;
+      padding-bottom: 8px;
+    `;
+
+    const inputContainerStyle = css`
+      position: relative;
+      display: flex;
+      align-items: center;
+      z-index: 0;
+    `;
+
+    const inputStyle = css`
+      width: 400px;
+      height: 36px;
+      border-radius: 4px;
+      padding-left: 12px;
+      font-size: 14px;
+      font-weight: normal;
+      &:placeholder {
+        color: #89989b;
+      }
+      &:focus {
+        outline: none;
+      }
+    `;
+
+    const errorIconStyle = css`
+      position: absolute;
+      right: 10px;
+      color: #cf4a22;
+    `;
+
+    const validIconStyle = css`
+      position: absolute;
+      right: 10px;
+      color: #13aa52;
+    `;
+
+    const optionalStyle = css`
+      position: absolute;
+      right: 12px;
+      color: #5d6c74;
+      font-size: 12px;
+      font-style: italic;
+      font-weight: normal;
+    `;
+
+    const errorMessageStyle = css`
+      color: #cf4a22;
+      font-size: 14px;
+      height: 20px;
+      padding-top: 4px;
+      font-weight: normal;
+    `;
+
+    return (
+      <div className={className}>
```

Comment:
Is there a reason that we're using a div as the container element? I think we should leave the container element as the label and apply the supplied className prop to the label element.

Reply:
Nope, no particular reason. I've reworked this so that the container element is the label.